src/mlpack/core/data/detect_file_type.cpp

```diff
@@ -0,0 +1,290 @@
+
+  for (arma::uword i = 0; i < nUse; ++i)
+  {
+    const unsigned char val = dataMem[i];
+    if ((val <= 8) || (val >= 123))
+    {
+      hasBinary = true;
+      break;
+    }  // The range checking can be made more elaborate.
+
+    if ((val == '(') || (val == ')'))
+    {
+      hasBracket = true;
+    }
+    if (val == ',')
+    {
+      hasComma = true;
+    }
+  }
+
+  delete[] dataMem;
+
+  if (hasBinary)
+    return arma::raw_binary;
+
+  if (hasComma && (hasBracket == false))
+    return arma::csv_ascii;
+
+  return arma::raw_ascii;
+}
+
+/**
+ * Attempt to auto-detect the type of a file given its extension, and by
+ * inspecting the parts of the file to disambiguate between types when
+ * necessary.  (For instance, a .csv file could be delimited by spaces, commas,
+ * or tabs.)  This is meant to be used during loading.
+ *
+ * @param stream Opened file stream to look into for autodetection.
+ * @param filename Name of the file.
+ * @return The detected file type.
+ */
+arma::file_type AutoDetect(std::fstream& stream,
+                           const std::string& filename)
+{
+  // Get the extension.
+  std::string extension = Extension(filename);
+  arma::file_type detectedLoadType = arma::file_type_unknown;
+
+  if (extension == "csv" || extension == "tsv")
+  {
+    detectedLoadType = GuessFileType(stream);
+    if (detectedLoadType == arma::csv_ascii)
+    {
+      if (extension == "tsv")
+        Log::Warn << "'" << filename << "' is comma-separated, not "
+            "tab-separated!" << std::endl;
+    }
+    else if (detectedLoadType == arma::raw_ascii) // .csv file can be tsv.
+    {
+      if (extension == "csv")
+      {
+        // We should issue a warning, but we don't want to issue the warning if
+        // there is only one column in the CSV (since there will be no commas
+        // anyway, and it will be detected as arma::raw_ascii).
+        const std::streampos pos = stream.tellg();
+        std::string line;
+        std::getline(stream, line, '\n');
+        boost::trim(line);
+
+        // Reset stream position.
+        stream.seekg(pos);
+
+        // If there are no spaces or whitespace in the line, then we shouldn't
+        // print the warning.
+        if ((line.find(' ') != std::string::npos) ||
+            (line.find('\t') != std::string::npos))
+        {
+          Log::Warn << "'" << filename << "' is not a standard csv file."
+              << std::endl;
+        }
+      }
+    }
+    else
+    {
+      detectedLoadType = arma::file_type_unknown;
+    }
+  }
+  else if (extension == "txt")
+  {
+    // This could be raw ASCII or Armadillo ASCII (ASCII with size header).
+    // We'll let Armadillo do its guessing (although we have to check if it is
+    // arma_ascii ourselves) and see what we come up with.
+
+    // This is taken from load_auto_detect() in diskio_meat.hpp
+    const std::string ARMA_MAT_TXT = "ARMA_MAT_TXT";
+    // char* rawHeader = new char[ARMA_MAT_TXT.length() + 1];
```

Comment:
Do you like to keep the line for reference?

Reply:
Good point, removed in 065ae8b9.